nemo_skills/code_execution/utils.py

```diff
@@ -87,23 +87,28 @@ def extract_code_output(generation: str, code_output_begin: str, code_output_end
     return _extract_between_separators(generation, [code_output_begin, code_output_end], extract_all)
 
 
-def extract_code_block(text: str, languages=None) -> str:
+def extract_code_block(text: str, languages=None, extract_mode: str = "last") -> str:
     if languages is None:
         languages = [""]
     for language in languages:
-        match = re.search(rf"```{language}\s*\n?(.*?)\n?```", text, re.DOTALL)
-        if match:
-            return match.group(1).strip()
+        matches = re.findall(rf"```{language}\s*\n?(.*?)\n?```", text, re.DOTALL)
+        if matches:
+            idx = 0 if extract_mode == "first" else -1
+            return matches[idx].strip()
     return ""
 
 
-def clean_formal_generation(generation: str, final_answer_key: str = "**FINAL ANSWER**") -> str:
+def clean_formal_generation(
+    generation: str,
+    final_answer_key: str = "**FINAL ANSWER**",
+    code_block_from: str = "first",
```

Comment:
```suggestion
    code_block_from: str = "last",
```
using last by default makes more sense I guess?

Reply:
I went with first to match the previous re.search behavior and passed through last just for lean

Reply:
Merging with the change, but if you or @gwarmstrong think the default should be first, let's update it in another pr